File: src/exoticatechnologies/modifications/upgrades/methods/ChipMethod.java
package exoticatechnologies.modifications.upgrades.methods;

import com.fs.starfarer.api.Global;
import com.fs.starfarer.api.campaign.CampaignFleetAPI;
import com.fs.starfarer.api.campaign.CargoAPI;
import com.fs.starfarer.api.campaign.CargoStackAPI;
import com.fs.starfarer.api.campaign.econ.MarketAPI;
import com.fs.starfarer.api.fleet.FleetMemberAPI;
import com.fs.starfarer.api.impl.campaign.ids.Commodities;
import com.fs.starfarer.api.util.Misc;
import exoticatechnologies.hullmods.ExoticaTechHM;
import exoticatechnologies.modifications.ShipModifications;
import exoticatechnologies.modifications.upgrades.Upgrade;
import exoticatechnologies.modifications.upgrades.UpgradeSpecialItemPlugin;
import exoticatechnologies.util.StringUtils;
import exoticatechnologies.util.Utilities;
import lombok.Getter;
import lombok.Setter;

import java.util.HashMap;
import java.util.Map;

public class ChipMethod implements UpgradeMethod {
    @Getter
    @Setter
    public CargoStackAPI upgradeChipStack = null;

    @Override
    public String getOptionText(FleetMemberAPI fm, ShipModifications mods, Upgrade upgrade, MarketAPI market) {
        return StringUtils.getString("UpgradeMethods", "ChipOption");
    }

    @Override
    public String getOptionTooltip(FleetMemberAPI fm, ShipModifications es, Upgrade upgrade, MarketAPI market) {
        return StringUtils.getString("UpgradeMethods", "ChipOptionTooltip");
    }

    @Override
    public boolean canUse(FleetMemberAPI fm, ShipModifications mods, Upgrade upgrade, MarketAPI market) {
        if (upgradeChipStack != null) {
            int creditCost = getCreditCost(fm, mods, upgrade, upgradeChipStack);
            return Global.getSector().getPlayerFleet().getCargo().getCredits().get() >= creditCost;
        }
        return Utilities.hasUpgradeChip(fm.getFleetData().getFleet().getCargo(), upgrade.getKey());
    }

    @Override
    public boolean canShow(FleetMemberAPI fm, ShipModifications mods, Upgrade upgrade, MarketAPI market) {
        return Utilities.hasUpgradeChip(fm.getFleetData().getFleet().getCargo(), upgrade.getKey());
    }

    @Override
    public String apply(FleetMemberAPI fm, ShipModifications mods, Upgrade upgrade, MarketAPI market) {
        if (upgradeChipStack == null) {
            throw new IllegalArgumentException("Missing chip stack.");
        }

        CampaignFleetAPI fleet = Global.getSector().getPlayerFleet();

        UpgradeSpecialItemPlugin stackPlugin = (UpgradeSpecialItemPlugin) upgradeChipStack.getPlugin();
        int creditCost = getCreditCost(fm, mods, upgrade, upgradeChipStack);

        fleet.getCargo().getCredits().subtract(creditCost);
        Utilities.takeItem(upgradeChipStack);

        mods.putUpgrade(upgrade, stackPlugin.getUpgradeLevel());
        mods.save(fm);
        ExoticaTechHM.addToFleetMember(fm);

        Global.getSoundPlayer().playUISound("ui_char_increase_skill_new", 1f, 1f);

        return StringUtils.getTranslation("UpgradesDialog", "UpgradePerformedSuccessfully")
                .format("name", upgrade.getName())
                .format("level", mods.getUpgrade(upgrade))
                .toString();
    }

    /**
     * returns the most valuable and still usable chip for the ship and its bandwidth.
     * @param fm the ship
     * @param mods the mods
     * @param upgrade the upgrade
     * @return the stack
     */
    public static CargoStackAPI getDesiredChip(FleetMemberAPI fm, ShipModifications mods, Upgrade upgrade) {
        return Utilities.getUpgradeChip(fm.getFleetData().getFleet().getCargo(), fm, mods, upgrade);
    }

    /**
     * Sums up the floats in the map.
     *
     * @param resourceCosts
     * @return The sum.
     */
    public static int getCreditCostForResources(Map<String, Integer> resourceCosts) {
        float creditCost = 0;

        for (Map.Entry<String, Integer> resourceCost : resourceCosts.entrySet()) {
            creditCost += Utilities.getItemPrice(resourceCost.getKey()) * resourceCost.getValue();
        }
        return (int) creditCost;
    }

    public static int getCreditCost(FleetMemberAPI fm, ShipModifications mods, Upgrade upgrade, CargoStackAPI stack) {
        if (stack != null) {
            UpgradeSpecialItemPlugin plugin = (UpgradeSpecialItemPlugin) stack.getPlugin();

            float resourceCreditCost = getCreditCostForResources(upgrade.getResourceCosts(fm, plugin.getUpgradeLevel()));
            int creditCost = (int) (resourceCreditCost * 0.33);

            return creditCost;
        }
        return 0;
    }

    @Override
    public Map<String, Float> getResourceCostMap(FleetMemberAPI fm, ShipModifications mods, Upgrade upgrade, MarketAPI market, boolean hovered) {
        Map<String, Float> resourceCosts = new HashMap<>();

        if (hovered) {
            if (upgradeChipStack == null) {
                String resourceName = StringUtils.getTranslation("ShipListDialog", "ChipName")
                        .format("name", upgrade.getName())
                        .toString();

                resourceCosts.put(String.format("&%s", resourceName), 1f);

                return resourceCosts;
            }

            resourceCosts.put(Commodities.CREDITS, (float) getCreditCost(fm, mods, upgrade, upgradeChipStack));
            resourceCosts.put(Utilities.formatSpecialItem(upgradeChipStack.getSpecialDataIfSpecial()), 1f);
        }

        return resourceCosts;
    }

    @Override
    public boolean usesBandwidth() {
        return true;
    }

    @Override
    public boolean usesLevel() {
        return true;
    }
}

File: src/exoticatechnologies/ui/impl/shop/upgrades/methods/ChipMethod.kt
package exoticatechnologies.ui.impl.shop.upgrades.methods

import com.fs.starfarer.api.Global
import com.fs.starfarer.api.campaign.CargoStackAPI
import com.fs.starfarer.api.campaign.econ.MarketAPI
import com.fs.starfarer.api.combat.ShipVariantAPI
import com.fs.starfarer.api.fleet.FleetMemberAPI
import com.fs.starfarer.api.impl.campaign.ids.Commodities
import exoticatechnologies.hullmods.ExoticaTechHM
import exoticatechnologies.modifications.ShipModLoader
import exoticatechnologies.modifications.ShipModifications
import exoticatechnologies.modifications.upgrades.Upgrade
import exoticatechnologies.modifications.upgrades.UpgradeSpecialItemPlugin
import exoticatechnologies.ui.impl.shop.upgrades.chips.UpgradeChipSearcher
import exoticatechnologies.util.StringUtils
import exoticatechnologies.util.Utilities

class ChipMethod : DefaultUpgradeMethod() {
    override val key = "chip"
    var upgradeChipStack: CargoStackAPI? = null
    override fun getOptionText(
        member: FleetMemberAPI,
        mods: ShipModifications,
        upgrade: Upgrade,
        market: MarketAPI?
    ): String {
        return StringUtils.getString("UpgradeMethods", "ChipOption")
    }

    override fun getOptionTooltip(
        member: FleetMemberAPI,
        mods: ShipModifications,
        upgrade: Upgrade,
        market: MarketAPI?
    ): String {
        return StringUtils.getString("UpgradeMethods", "ChipOptionTooltip")
    }

    override fun canUse(member: FleetMemberAPI, mods: ShipModifications, upgrade: Upgrade, market: MarketAPI?): Boolean {
        market ?: return false

        if (upgradeChipStack != null) {
            val creditCost = getCreditCost(member, mods, upgrade, upgradeChipStack)
            return Global.getSector().playerFleet.cargo.credits.get() >= creditCost
        }

        return (UpgradeChipSearcher().getChips(Global.getSector().playerFleet.cargo, member, mods, upgrade).isNotEmpty()
                && super.canUse(member, mods, upgrade, market))
    }

    override fun canShow(member: FleetMemberAPI, mods: ShipModifications, upgrade: Upgrade, market: MarketAPI?): Boolean {
        return Utilities.hasUpgradeChip(Global.getSector().playerFleet.cargo, upgrade.key)
    }

    override fun apply(
        member: FleetMemberAPI,
        variant: ShipVariantAPI,
        mods: ShipModifications,
        upgrade: Upgrade,
        market: MarketAPI?
    ): String {
        requireNotNull(upgradeChipStack) { "Missing chip stack." }

        val fleet = Global.getSector().playerFleet
        val stackPlugin = upgradeChipStack!!.plugin as UpgradeSpecialItemPlugin
        val creditCost = getCreditCost(member, mods, upgrade, upgradeChipStack)
        fleet.cargo.credits.subtract(creditCost.toFloat())
        Utilities.takeItem(upgradeChipStack)

        mods.putUpgrade(upgrade, stackPlugin.upgradeLevel)
        ShipModLoader.set(member, variant, mods)
        ExoticaTechHM.addToFleetMember(member, variant)

        return StringUtils.getTranslation("UpgradesDialog", "UpgradePerformedSuccessfully")
            .format("name", upgrade.name)
            .format("level", mods.getUpgrade(upgrade))
            .toString()
    }

    override fun getResourceCostMap(
        member: FleetMemberAPI,
        mods: ShipModifications,
        upgrade: Upgrade,
        market: MarketAPI?,
        hovered: Boolean
    ): Map<String, Float> {
        val resourceCosts: MutableMap<String, Float> = HashMap()
        if (hovered) {
            if (upgradeChipStack == null) {
                val resourceName = StringUtils.getTranslation("ShipListDialog", "ChipName")
                    .format("name", upgrade.name)
                    .toString()
                resourceCosts[String.format("&%s", resourceName)] = 1f
                return resourceCosts
            }
            resourceCosts[Commodities.CREDITS] = getCreditCost(member, mods, upgrade, upgradeChipStack).toFloat()
            resourceCosts[Utilities.formatSpecialItem(upgradeChipStack!!.specialDataIfSpecial)] =
                1f
        }
        return resourceCosts
    }

    companion object {
        /**
         * returns the most valuable and still usable chip for the ship and its bandwidth.
         * @param fm the ship
         * @param mods the mods
         * @param upgrade the upgrade
         * @return the stack
         */
        fun getDesiredChip(fm: FleetMemberAPI, mods: ShipModifications, upgrade: Upgrade): CargoStackAPI? {
            return Utilities.getUpgradeChip(Global.getSector().playerFleet.cargo, fm, mods, upgrade)
        }

        /**
         * Sums up the floats in the map.
         *
         * @param resourceCosts
         * @return The sum.
         */
        fun getCreditCostForResources(resourceCosts: Map<String, Int>): Int {
            var creditCost = 0f
            for ((key, value) in resourceCosts) {
                creditCost += Utilities.getItemPrice(key) * value
            }
            return creditCost.toInt()
        }

        fun getCreditCost(fm: FleetMemberAPI, mods: ShipModifications, upgrade: Upgrade, stack: CargoStackAPI?): Int {
            if (stack != null) {
                val plugin = stack.plugin as UpgradeSpecialItemPlugin
                val resourceCreditCost =
                    getCreditCostForResources(
                        upgrade.getResourceCosts(fm, plugin.upgradeLevel)
                    ).toFloat()
                return (resourceCreditCost * 0.33).toInt()
            }
            return 0
        }
    }
}